src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php

```diff
@@ -264,6 +264,26 @@ private function addFragmentsSection(ArrayNodeDefinition $rootNode)
     private function addProfilerSection(ArrayNodeDefinition $rootNode)
     {
         $rootNode
+            ->beforeNormalization()
+                ->ifTrue(function ($v) { return 'file:' !== substr($v['dsn'], 0, 5); })
+                ->then(function ($v) {
+                    @trigger_error('The profiler.dsn configuration key must start with "file:" because all the storages except the filesystem are deprecated since version 2.8 and will be removed in 3.0.', E_USER_DEPRECATED);
+
+                    return $v;
+                })
+                ->ifTrue(function ($v) { return array_key_exists('username', $v); })
+                ->then(function ($v) {
+                    @trigger_error('The profiler.username configuration key is deprecated since version 2.8 and will be removed in 3.0.', E_USER_DEPRECATED);
```

Comment:
given that you need to trigger it all the time when ``username`` is configured, and you don't need to access the other fields, it would be better to put this ``beforeNormalization`` directly on the ``username`` node (and then, you can use ``->always()`` for the condition as the processing will only normalize the value if it is configured)